src/Dialect/ONNX/IndexExpr.hpp

```diff
@@ -0,0 +1,348 @@
+
+  // IndexExpr basic builders.
+  IndexExpr CreateUndefinedIndex();
+  IndexExpr CreateQuestionmarkIndex();
+  IndexExpr CreateLiteralIndex(int64_t val);
+  IndexExpr CreateDimIndex(Value val);
+  IndexExpr CreateSymbolIndex(Value val);
+
+  // Scan a memref shape at index to generate an IndexExpr, typically used for
+  // dimensions. Generate a literal when the memref dimension is known at
+  // compile time.
+  IndexExpr CreateDimIndexFromMemref(
+      Value memref, ArrayRef<int64_t> memrefShape, int index);
+  // Consider an op with operand "arrayOperand". We find this operand's defining
+  // op: if it contains a literal at position "index", we generate an literal
+  // IndexExpr; if its a tensor/memref, we load this value. If the index is out
+  // of bound, we return an undefine IndexExpr.
+  IndexExpr CreateSymbolIndexFromArrayAtIndex(
+      Operation *op, Value arrayOperand, uint64_t index);
+  // Same as above, but return "defaultLitteral" when there are no defining op
+  // or the index is out of bound.
+  IndexExpr CreateSymbolIndexFromArrayAtIndex(Operation *op, Value arrayOperand,
+      uint64_t index, int64_t defaultLiteral);
+
+  // Additional builder for repurposing IndexExpr from parent context.
+  IndexExpr CreateSymbolIndexFromParentContext(IndexExpr &parentIndexExpr);
+
+  // Actions for AffineExpr.
+  int AddDim(Value value);
+  int AddSymbol(Value value);
+
+  // Querries.
+  bool IsShapeInferencePass() const { return !rewriter; }
+
+  // Getters.
+  void GetDimAndSymbolList(SmallVectorImpl<Value> &list) const;
+  int GetDimSize() const { return dims.size(); }
+  int GetSymbolSize() const { return symbols.size(); }
+  ConversionPatternRewriter &GetRewriter() const;
+  Location GetLocation() const { return loc; }
+
+  // Static helper functions.
+  bool static AreAllLiteral(SmallVectorImpl<IndexExpr> &list);
+  bool static AreAllAffine(SmallVectorImpl<IndexExpr> &list);
+  void static GetOutputDimsForType(SmallVectorImpl<IndexExpr> &outputIndices,
+      SmallVectorImpl<int64_t> &outputDims);
+
+private:
+  SmallVector<Value, 4> dims;
+  SmallVector<Value, 4> symbols;
+  ConversionPatternRewriter *rewriter;
+  Location loc;
+  IndexExprContext *parentContext;
+};
+
+class IndexExpr {
+public:
+  friend class IndexExprContext;
+
+  IndexExpr();
+
+  // Shape inference querries.
+  bool IsDefined() const;
+  bool IsUndefined() const { return !IsDefined(); }
+  bool IsLiteral() const;
+  bool IsQuestionmark() const;
+  bool IsAffine() const;
+  bool IsSymbol() const;
+  bool IsDim() const;
+  bool IsShapeInferencePass() const;
+  bool HasContext() const;
+  bool HasAffineExpr() const;
+  bool HasValue() const;
+
+  // Getters.
+  int64_t GetLiteral() const;
+  AffineExpr GetAffineExpr();
+  Value GetValue();
+  IndexExprContext &GetContext() const;
+  IndexExprContext *GetContextPtr() const;
+  Location GetLocation() const;
+
+  // Possibly Affine Operations.
+  IndexExpr &Add(IndexExpr &a, IndexExpr &b);
```

Comment:
If all the operands are provided, are these operation static to the class? I assume that  we will define operator  for them later.

Reply:
No, it writes into the *this object, and then returns the *this.

That way you can have chains: 

```
newObj.add(a, b).divby(c)
```

First newObj becomes `a+b` and then that result is divided by `c`.

I did not want to use dynamic memory to alloc/free on demand...